ProjFS.Mac/PrjFSKext/VnodeCache.cpp

```diff
@@ -0,0 +1,313 @@
+#include <string.h>
+#include "Locks.hpp"
+#include "VnodeCache.hpp"
+#include "Memory.hpp"
+#include "KextLog.hpp"
+#include "../PrjFSKext/public/PrjFSCommon.h"
+
+#include "VnodeCachePrivate.hpp"
+
+#ifdef KEXT_UNIT_TESTING
+#include "VnodeCacheTestable.hpp"
+#endif
+
+KEXT_STATIC_INLINE void InvalidateCache_ExclusiveLocked();
+KEXT_STATIC_INLINE uintptr_t HashVnode(vnode_t _Nonnull vnode);
+
+KEXT_STATIC bool TryGetVnodeRootFromCache(
+    vnode_t _Nonnull vnode,
+    uintptr_t vnodeHash,
+    uint32_t vnodeVid,
+    /* out parameters */
+    VirtualizationRootHandle& rootHandle);
+
+KEXT_STATIC void UpdateCacheForVnode(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vfs_context_t _Nonnull context,
+    vnode_t _Nonnull vnode,
+    uintptr_t vnodeHash,
+    uint32_t vnodeVid,
+    bool invalidateEntry,
+    /* out parameters */
+    VirtualizationRootHandle& rootHandle);
+
+KEXT_STATIC bool TryFindVnodeIndex_Locked(
+    vnode_t _Nonnull vnode,
+    uintptr_t vnodeHash,
+    /* out parameters */
+    uintptr_t& vnodeIndex);
+
+KEXT_STATIC bool TryUpdateEntryToLatest_ExclusiveLocked(
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vnode_t _Nonnull vnode,
+    vfs_context_t _Nonnull context,
+    uintptr_t vnodeHash,
+    uint32_t vnodeVid,
+    bool invalidateEntry,
+    /* out parameters */
+    VirtualizationRootHandle& rootHandle);
+
+KEXT_STATIC uint32_t s_entriesCapacity;
+KEXT_STATIC VnodeCacheEntry* s_entries;
+static RWLock s_entriesLock;
+static const uint32_t MinEntriesCapacity = 0x040000; //  4 MB (assuming 16 bytes per VnodeCacheEntry)
+static const uint32_t MaxEntriesCapacity = 0x400000; // 64 MB (assuming 16 bytes per VnodeCacheEntry)
+
+kern_return_t VnodeCache_Init()
+{
+    if (RWLock_IsValid(s_entriesLock))
+    {
+        return KERN_FAILURE;
+    }
+    
+    s_entriesLock = RWLock_Alloc();
+    if (!RWLock_IsValid(s_entriesLock))
+    {
+        return KERN_FAILURE;
+    }
+
+    s_entriesCapacity = Clamp(desiredvnodes * 2u, MinEntriesCapacity, MaxEntriesCapacity);
+    
+    s_entries = Memory_AllocArray<VnodeCacheEntry>(s_entriesCapacity);
+    if (nullptr == s_entries)
+    {
+        s_entriesCapacity = 0;
+        return KERN_RESOURCE_SHORTAGE;
+    }
+    
+    VnodeCache_InvalidateCache(nullptr);
+    
+    PerfTracing_RecordSample(PrjFSPerfCounter_CacheCapacity, 0, s_entriesCapacity);
+    
+    return KERN_SUCCESS;
+}
+
+void VnodeCache_Cleanup()
```

Comment:
Do we need to worry about thread safety with this function?

Reply:
`VirtualizationRoots_Cleanup` is called [after `AbortAllOutstandingEvents`](https://github.com/wilbaker/VFSForGit/blob/fb66e31b9a1e02e25d62a462fd68862ae3dc2d9b/ProjFS.Mac/PrjFSKext/KauthHandler.cpp#L208) and so there should no longer be any threads trying to access it.

This is the same approach used in `VirtualizationRoots_Cleanup` and so far that's been okay.  However, your comment made me realize that `VirtualizationRoots_Cleanup` is a bit more cautious:

```
    if (RWLock_IsValid(s_virtualizationRootsLock))
    {
        RWLock_FreeMemory(&s_virtualizationRootsLock);
        return KERN_SUCCESS;
    }
```

And I can add that same check here.